spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,549 @@
+# In overall metrics (`f1`, `accuracy`) our model appears to perform well!
+#
+# However, we emphasize here that more often than not, we're interested in performance for application-critical subsets, or _slices_.
+#
+# Let's perform an [`error_analysis`](https://snorkel.readthedocs.io/en/redux/source/snorkel.analysis.html#module-snorkel.analysis.error_analysis) to see where our model makes mistakes.
+# We'll collect the predictions from the model and visualize examples in specific error buckets.
+
+# %%
+from snorkel.analysis.error_analysis import get_label_buckets
+from snorkel.analysis.utils import probs_to_preds
+
+outputs = model.predict(dl_valid, return_preds=True)
+buckets = get_label_buckets(
+    outputs["golds"]["spam_task"], outputs["preds"]["spam_task"]
+)
+
+# %% [markdown]
+# For application purposes, we might care especially about false negatives (true label was `1`, but model predicted `0`)—we want to minimize the amount of spam that our users are seeing!
+
+# %%
+df_valid[["text", "label"]].iloc[buckets[(1, 0)]]
+
+# %% [markdown]
+# We notice that we're mis-classifying particularly some comments with shortened urls (e.g. `bit.ly/...`) — these links could redirect us to potentially dangerous websites, and we don't want our users to click them!
+
+# %% [markdown]
+# ## 3. Monitor data slices
+#
+# We leverage *slicing functions* (SFs)— an abstraction that shares syntax with *labeling functions*, which you should already be familiar with! (If not, please see the [intro tutorial](https://github.com/snorkel-team/snorkel-tutorials/blob/master/spam/01_spam_tutorial.ipynb).) A key difference: whereas labeling functions output labels, slicing functions output binary _masks_ indicating whether an example is in the slice or not.
+
+# %% [markdown]
+# In the following cells, we define a slicing function that identifies these shortened links the spam dataset.
+# To do so, we write a regex that checks for the commonly-used `.ly` extension.
+#
+# You'll notice that the slicing function is noisily defined—SFs are often heuristics to quickly measure performance over important subsets of the data.
+
+# %%
+import re
+from snorkel.slicing.sf import slicing_function
+
+
+@slicing_function()
+def short_link(x):
+    """Spam comments post links to other channels."""
+    return bool(re.search(r"\w+\.ly", x.text))
+
+
+sfs = [short_link]
+slice_names = [sf.name for sf in sfs]
+
+# %% [markdown]
+# For our $n$ examples and $k$ slices in each split, we apply the SF to our data to create an $n \times k$ matrix. (So far, $k=1$).
+
+# %%
+from snorkel.slicing.apply import PandasSFApplier
+
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %% [markdown]
+# With a utility function from `snorkel.analysis.monitor`, we can visualize examples belonging to this slice in a `pandas.DataFrame`.
+
+# %%
+from snorkel.analysis.monitor import slice_df
+
+short_link_df = slice_df(
```

Comment:
hm ok don't love this interface, way too much repeated passing between this and SFApplier. let's go for cleanliness over saving repeated computation here. i'd suggest something that looks like.

```python
train_slicer = PandasSlicer(df_train)
df_train_short_link = train_slicer.slice(short_link)
```

PandasSlicer.slice will just call PandasSFApplier under the hood

Reply:
the reason we pass df_train to init (backwards from applier) is because i can see train_slicer being reused for multiple sfs

Reply:
cool! i like the idea of cleanliness >> computation 

new api:
```
slicer = PandasSlicer(df_train, sfs, slice_names)
slicer.slice("short_link")
```